File: backend/src/main/java/com/hrms/application/ai/service/AIRecruitmentService.java
package com.hrms.application.ai.service;

import com.hrms.api.recruitment.dto.ai.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

/**
 * Facade for AI-powered recruitment capabilities.
 *
 * <p>Delegates to three focused sub-services:
 * <ul>
 *   <li>{@link ResumeParserService} — resume parsing and text extraction</li>
 *   <li>{@link CandidateMatchingService} — scoring, ranking, and JD generation</li>
 *   <li>{@link InterviewGenerationService} — question generation and feedback synthesis</li>
 * </ul>
 *
 * <p>Preserves the original public API so that all existing callers (controllers, tests)
 * require no changes.
 */
@Service
@RequiredArgsConstructor
@Transactional
public class AIRecruitmentService {

    private final ResumeParserService resumeParser;
    private final CandidateMatchingService candidateMatching;
    private final InterviewGenerationService interviewGeneration;

    // ==================== RESUME PARSING ====================

    public ResumeParseResponse parseResume(String resumeText) {
        return resumeParser.parseResume(resumeText);
    }

    public ResumeParseResponse parseResumeFromUrl(String resumeUrl) {
        return resumeParser.parseResumeFromUrl(resumeUrl);
    }

    public ResumeParseResponse parseResumeFromFile(byte[] fileBytes, String fileName) {
        return resumeParser.parseResumeFromFile(fileBytes, fileName);
    }

    public ResumeParseResponse parseResumeFromBase64(String base64Content, String fileName) {
        return resumeParser.parseResumeFromBase64(base64Content, fileName);
    }

    // ==================== CANDIDATE MATCHING & SCREENING ====================

    @Transactional(readOnly = true)
    public CandidateMatchResponse calculateMatchScore(UUID candidateId, UUID jobOpeningId) {
        return candidateMatching.calculateMatchScore(candidateId, jobOpeningId);
    }

    @Transactional(readOnly = true)
    public CandidateScreeningSummaryResponse generateScreeningSummary(
            UUID candidateId, UUID jobOpeningId, String context) {
        return candidateMatching.generateScreeningSummary(candidateId, jobOpeningId, context);
    }

    public List<CandidateMatchResponse> rankCandidatesForJob(UUID jobOpeningId) {
        return candidateMatching.rankCandidatesForJob(jobOpeningId);
    }

    // ==================== JOB DESCRIPTION GENERATION ====================

    @Transactional(readOnly = true)
    public JobDescriptionResponse generateJobDescription(JobDescriptionRequest request) {
        return candidateMatching.generateJobDescription(request);
    }

    // ==================== INTERVIEW QUESTIONS ====================

    @Transactional(readOnly = true)
    public InterviewQuestionsResponse generateInterviewQuestions(UUID jobOpeningId, UUID candidateId) {
        return interviewGeneration.generateInterviewQuestions(jobOpeningId, candidateId);
    }

    @Transactional(readOnly = true)
    public InterviewQuestionsResponse generateInterviewQuestions(UUID jobOpeningId) {
        return interviewGeneration.generateInterviewQuestions(jobOpeningId);
    }

    // ==================== FEEDBACK SYNTHESIS ====================

    public FeedbackSynthesisResponse synthesizeInterviewFeedback(UUID candidateId, UUID jobOpeningId) {
        return interviewGeneration.synthesizeInterviewFeedback(candidateId, jobOpeningId);
    }
}

File: frontend/lib/services/hire/ai-recruitment.service.ts
import {apiClient} from '../../api/client';
import {
  CandidateMatchResponse,
  CandidateScreeningSummaryRequest,
  CandidateScreeningSummaryResponse,
  FeedbackSynthesisRequest,
  FeedbackSynthesisResponse,
  InterviewQuestionsResponse,
  JobDescriptionRequest,
  JobDescriptionResponse,
  ResumeParseRequest,
  ResumeParseResponse,
} from '../../types/hire/ai-recruitment';

class AIRecruitmentService {
  // ==================== Resume Parsing ====================

  async parseResume(request: ResumeParseRequest): Promise<ResumeParseResponse> {
    const response = await apiClient.post<ResumeParseResponse>(
      '/recruitment/ai/parse-resume',
      request
    );
    return response.data;
  }

  /**
   * Parse resume from a multipart file upload (PDF, DOCX, etc.)
   * Uses the /parse-resume/upload endpoint with FormData.
   */
  async parseResumeFromUpload(file: File): Promise<ResumeParseResponse> {
    const formData = new FormData();
    formData.append('file', file);
    const response = await apiClient.post<ResumeParseResponse>(
      '/recruitment/ai/parse-resume/upload',
      formData,
      {headers: {'Content-Type': 'multipart/form-data'}}
    );
    return response.data;
  }

  // ==================== Candidate Matching ====================

  async calculateMatchScore(
    candidateId: string,
    jobOpeningId: string
  ): Promise<CandidateMatchResponse> {
    const response = await apiClient.post<CandidateMatchResponse>(
      '/recruitment/ai/match-score',
      null,
      {params: {candidateId, jobOpeningId}}
    );
    return response.data;
  }

  async rankCandidatesForJob(jobOpeningId: string): Promise<CandidateMatchResponse[]> {
    const response = await apiClient.get<CandidateMatchResponse[]>(
      `/recruitment/ai/ranked-candidates/${jobOpeningId}`
    );
    return response.data;
  }

  // ==================== Screening Summary ====================

  async generateScreeningSummary(
    request: CandidateScreeningSummaryRequest
  ): Promise<CandidateScreeningSummaryResponse> {
    const response = await apiClient.post<CandidateScreeningSummaryResponse>(
      '/recruitment/ai/screening-summary',
      request
    );
    return response.data;
  }

  // ==================== Job Description ====================

  async generateJobDescription(
    request: JobDescriptionRequest
  ): Promise<JobDescriptionResponse> {
    const response = await apiClient.post<JobDescriptionResponse>(
      '/recruitment/ai/generate-job-description',
      request
    );
    return response.data;
  }

  // ==================== Interview Questions ====================

  async generateInterviewQuestions(
    jobOpeningId: string,
    candidateId?: string
  ): Promise<InterviewQuestionsResponse> {
    const params: Record<string, string> = {};
    if (candidateId) {
      params.candidateId = candidateId;
    }
    const response = await apiClient.get<InterviewQuestionsResponse>(
      `/recruitment/ai/interview-questions/${jobOpeningId}`,
      {params}
    );
    return response.data;
  }

  // ==================== Feedback Synthesis ====================

  async synthesizeFeedback(
    request: FeedbackSynthesisRequest
  ): Promise<FeedbackSynthesisResponse> {
    const response = await apiClient.post<FeedbackSynthesisResponse>(
      '/recruitment/ai/synthesize-feedback',
      request
    );
    return response.data;
  }
}

export const aiRecruitmentService = new AIRecruitmentService();
